week1.md

```diff
@@ -9,17 +9,172 @@ these things yet):
+  the users participating in the conversation. The other option is explained in
+  more detail below.
+
+### Relationships with attributes
+
+Sometimes, a rerlationship its self may need to have attributes. For example, if
+we want to keep track of a user's role in a conversation, and say if they are
+the owner or a regular participant of the conversation, then this is an
+attribute of the relationship its self. The way to handle this situation is to
+turn the relationship into a new entity with 1-to-1 relationships with both
+other entities. In this example, the new entity would be called
+`ConversationUser`. It would have a schema that looks like this:
+
+### ConversationUserSchema
+
+| Attribute      | Type                     | Optional? | Other Notes                  |
+| -------------- | ------------------------ | --------- | ---------------------------- |
+| \_id           | ObjectId                 | N         | Primary key, Unique          |
+| userId         | ObjectId                 | N         | References User.\_id         |
+| conversationId | ObjectId                 | N         | References Conversation.\_id |
+| role           | "owner" \| "participant" | N         |                              |
+
+## Task
+
+Create a database schema here which has all the entities which we need to implement
+the chat system. Hint: read the requirements of the system in the README and use
+that as a starting point for listing entities.
+
+### Messenger application schema
+
+## User schema
+
+| Attribute      | Type                     | Optional? | Other Notes                  |
+| -------------- | ------------------------ | --------- | ---------------------------- |
+| \_id           | ObjectId                 | N         | Primary key, Unique          |
+| email          | String                   | N         | Unique, email address       |
+| password 			 | String                   | N         | Encrypted, must contain special characters, numbers, and at least a capital letter  |
+| contactList    | Object 									| N         | Here will be saved all of the user's contacts|
+| role           | "owner" \| "list"				| N 				| 															| 
+| fullName 			 | String 									| Y 				| Optional full name of the user. Can be left empty for anonymity |
+
+## ContactList schema
+
+| Attribute      | Type                     | Optional? | Other Notes                  |
+| -------------- | ------------------------ | --------- | ---------------------------- |
+| \_id           | ObjectId                 | N         | Primary key, Unique          |
+| owner          | Object                   | N         | User who owns this contact list |
+| contacts 			 | List                   	| N         | List of contacts the user has  |
+| contactList    | Object 									| N         | Here will be saved all of the user's contacts|
+| role           | "list"	\| "owner"  			| N 				| 															| 
+
+## Contact scheme
```

Comment:
I see what you're going for here, and I like the thought process, but I've got a challenge for this design:

1. Consider this from both user's perspective:
   - Assume user A is a contact of user B, and vice-versa
   - User A's ContactList will have an entry for User B. User B's contact list will also have an entry for User A.
   - There are two possible representations under this schema - either we A) make sure the `Contact` entry is shared between User A and User B, or we keep two copies of the `messages` in sync.  This will be made more complicated by the fact that you have collections for both `sent` and `received` messages - which implies keeping 4 different arrays in sync for every conversation.

For the sake of simplicity, let's assume that in this messenger, every user is able to maintain a conversation with every other user (instead of each user having a list of contacts).  In other words, if any user accesses the `users` screen, they will see every user in the system (apart from themselves).